src/olympia/amo/middleware.py

```diff
@@ -341,3 +350,50 @@ def __call__(self, request):
         elif max_age_from_response is None:
             patch_cache_control(response, s_maxage=0)
         return response
+
+
+class TokenValidMiddleware:
+    """Middleware to check the FxA auth tokens haven't expired, and refresh if
+    necessary."""
+
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request):
+        if (
+            # We check the api token in WebTokenAuthentication
+            not getattr(request, 'is_api', False)
+            and not settings.USE_FAKE_FXA_AUTH
+            and settings.VERIFY_FXA_ACCESS_TOKEN_WEB
+            and not check_fxa_access_token_validity(
+                request.session.get('access_token_expiry')
+            )
+        ):
+            fxa_token_object = update_fxa_access_token(
+                request.session.get('user_token_pk'),
+                request.user,
+            )
+            if not fxa_token_object:
+                print(
+                    'User access token refresh failed; '
+                    'so redirect to login to FxA again'
+                )
+                return redirect_for_login(request)
+            request.session[
+                'access_token_expiry'
+            ] = fxa_token_object.access_token_expiry.timestamp()
+            request._fxatoken = fxa_token_object
+
+        response = self.get_response(request)
+
+        # update frontend cookie if we have a refreshed token.
+        # `request._fxatoken` could be set above, or in WebTokenAuthentication
+        if fxa_token_object := getattr(request, '_fxatoken', None):
+            token = generate_api_token(
```

Comment:
Maybe as a pre-requisite to all of this, we should switch to a proper session backend that stores data server-side. The reason AMO switched to signed cookies 10 years ago, before my time, was because most (all?) pages would create a session even if the user was not logged in (for various reasons, like dealing with CSRF), causing a lot of extra load on the database, which was the session backend at the time. But this is no longer the case now...

Reply:
Until this patch was rewritten to avoid a database call on every request it didn't have to rewrite the cookies when the access_token had expired... and a database session backend will have to make a database call on every request. 😬

Reply:
does it actually matter that the AMO session lasts forever* if we check validity against FxA?  *(not actually forever - until the fxa access token is no longer refreshable)

Reply:
server-side storage doesn't necessarily mean using the MySQL database though, we could use redis or more likely memcached+MySQL. But I see your point :)